README.md
# Prisma Upgrade [![npm](https://img.shields.io/npm/v/prisma-upgrade)](https://www.npmjs.com/package/prisma-upgrade)

Prisma Upgrade is a CLI tool to help Prisma 1 users using MySQL or Postgres to upgrade to Prisma 2+.

## Scope

We spent a lot of time on this tool and are happy with the results. It's well-tested and working as intended. It's not perfect though, so you may need to make some manual adjustments after upgrading to clean up your final Prisma 2+ schema. See [this issue](https://github.com/prisma/upgrade/issues/67) for more details.

We recommend you download the [Prisma VS Code extension](https://marketplace.visualstudio.com/items?itemName=Prisma.prisma) to help with your transition.

Note: You should always run the SQL generated by this tool on your test or staging databases before running it on production.

## Usage

```sh
$ npx prisma-upgrade
```

See [our documentation](https://www.prisma.io/docs/guides/upgrade-from-prisma-1/how-to-upgrade#prisma-upgrade-cli) for more information about how to upgrade your Prisma 1 datamodel to Prisma 2+.

## Features

This table reflects the _current_ feature set of the upgrade CLI and will be updated continuously. Read below for a more detailed explanation of each column. 

| Problem                                                                                                                                                                                                                                                                               | MySQL   | PostgreSQL | Prisma schema | Prisma 1 compatible |
|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------|------------|---------------|---------------------|
| _Database schema incompatibilities²_                                                                                                                                                                                                                                                  |         |            |               |                     |
| [Default values aren't represented in database](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#default-values-arent-represented-in-database) <!-- SetDefaultOp -->                                                             | Yes     | Yes        | Yes           | Yes                 |
| [Mismatching CUID length](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#mismatching-cuid-length) <!-- AlterIDsOp -->                                                                                                          | Yes     | Yes        | Yes           | Yes                 |
| [`@createdAt` isn't represented in database](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#createdat-isnt-represented-in-database)<!-- SetCreatedAtOp -->                                                                     | Yes     | Yes        | Yes           | Yes                 |
| [Inline 1-1 relations are recognized as 1-n (missing `UNIQUE` constraint)](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#inline-1-1-relations-are-recognized-as-1-n-missing-unique-constraint) <!-- AddUniqueConstraintOp --> | Yes     | Yes        | Yes           | Yes                 |
| [`Json` type is represented as `TEXT`](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#json-type-is-represented-as-text-in-database) <!-- SetJsonTypeOp -->                                                                     | Yes     | Yes        | Yes           | Yes                 |
| [Enum types are represented as `TEXT` in database](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#enums-are-represented-as-text-in-database) <!-- SetEnumTypeOp -->                                                            | Yes     | Yes        | Yes           | Yes                 |
| [All non-inline relations are recognized as m-n](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#all-non-inline-relations-are-recognized-as-m-n) <!-- MigrateHasManyOp -->                                                      | Not yet | Not yet    | Not yet       | No                  |
| [Scalar lists (arrays) are maintained with extra table](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#scalar-lists-arrays-are-maintained-with-extra-table) <!-- MigrateScalarListOp -->                                       | Not yet | Not yet    | Not yet       | No                  |
| _Prisma 2+ schema differences³_                                                                                                                                                                                                                                                       |         |            |               |                     |
| [`@updatedAt` isn't represented in database](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#updatedat-isnt-represented-in-database) <!-- upgrade() -->                                                                         | n/a     | n/a        | Yes           | Yes                 |
| [Generated CUIDs as ID values aren't represented in database](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-mysql#generated-cuids-as-id-values-arent-represented-in-database) <!-- upgrade() -->                                    | n/a     | n/a        | Yes           | Yes                 |
| Maintain required 1-1-relations                                                                                                                                                                                                                                                       | n/a     | n/a        | Yes           | Yes                 |
| Maintain order of models and fields                                                                                                                                                                                                                                                   | n/a     | n/a        | Not yet       | Yes                 |
| Maintain relation names                                                                                                                                                                                                                                                               | n/a     | n/a        | Not yet       | Yes                 |
| `@map` and `@@map`                                                                                                                                                                                                                                                                   | n/a     | n/a        | Yes           | Yes                 |
| Cascading deletes                                                                                                                                                                                                                                                                     | No      | No         | No            | No                  |
<!-- TODO: MigrateEnumListOp -->

² = fixed by executing SQL statements  
³ = fixed by making changes to Prisma 2+ schema

What do the columns mean?

- **MySQL**: Does the CLI generate the correct MySQL statements to solve the problem?
- **PostgreSQL**: Does the CLI generate correct PostgreSQL statements to solve the problem?
- **Prisma schema**: Does the final Prisma 2+ schema I get from the CLI reflect the right solution?
- **Prisma 1 compatible:** Does the SQL change to the schema maintain Prisma 1 compatibility?

## How Prisma Upgrade (Technically) Works

We parse your Prisma 1 datamodel and your Prisma 2+ schema and run both ASTs through a set of rules. These rules produce operations. The operations are printed into SQL commands for you to run on your database.

Prisma upgrade is idempotent, so you can run it as many times as you want and it will produce the same result each time. Prisma upgrade only shows you commands you still need to run, it does not show you commands you've already run.

You'll also notice that we never connect to your database, we simply look at your Prisma 1 files and your Prisma 2+ schema and generate from there!

## Development

To add a new test, follow the following steps:

1. Create a new folder with one of the following prefixes: `postgres-`, `mysql-`, `postgres1-` or `mysql1-`. Use `postgres-` and `mysql-` for Prisma `Prisma 1.34`, `postgres1` and `mysql1-` for versions before that.
2. Add a `prisma.yml` and `datamodel.graphql`
3. Run `node scripts/sqldump.js`. This should product a `dump.sql` file and a `schema.prisma`.
4. Add an `expected.sql`. This is the commands you expect to `prisma-upgrade` to guide you to do. Often, I first leave this blank, see what `prisma-upgrade` does and then tweak.
5. Add an `expected.prisma`. This is the final state the schema should be in. Often, I first leave this blank, see what `prisma-upgrade` does and then tweak.

## Tests

Testing during implementation of this tool consists of 2 parts: a Local SQL Dump and Running Tests

### Local SQL Dump

_Requirements:_ MySQL@5, Docker

Since it's cumbersome to run Prisma 1 in CI, we need to locally setup test cases first

### Setting up MySQL for examples

```
mysqladmin -h localhost -u root create prisma
mysql -h localhost -u root prisma < ./examples/mysql-ablog/dump.sql
mysqladmin -h localhost -u root drop prisma -f
```

## Security

If you have a security issue to report, please contact us at [security@prisma.io](mailto:security@prisma.io?subject=[GitHub]%20Prisma%202%20Security%20Report%20Upgrade)


History.md
# 0.0.44 / 2022-04-05

- chore: add mongodb upgrade docs link (#134)
- chore: error for mongodb (#130)
- docs(README): Update and resort feature table (#126)
- Update mentions of "Prisma 2" to "Prisma 2+"
- Document use of `db pull` instead of `introspect`
- Error early when provider === mongodb as it's not supported
- Create SECURITY.md

# 0.0.43 / 2021-03-05

- Fix prisma/upgrade for native types.
- Add development instructions.
- Fix sqldump script for mysqldump@8
- switch from @prisma/cli to prisma
- Add note about MySQL and Postgres

# 0.0.42 / 2020-09-10

- keep json & enum operations in non-breaking changes as well for mysql

# 0.0.41 / 2020-09-10

- move json and enum operations to the breaking changes list

# 0.0.40 / 2020-08-24

- build project for previous fix

# 0.0.39 / 2020-08-24

- fix label in cli

# 0.0.38 / 2020-08-21

- migrate postgres json and enum to breaking changes. closes #73 #74
- char(25) => char(30)

# 0.0.37 / 2020-08-14

- fix 1-to-1 table relations for postgres and mysql

# 0.0.36 / 2020-08-13

- maintain relationship names from Prisma 1. Closes #25

# 0.0.35 / 2020-07-28

- support nullable lists in Postgres, even though Prisma 2 doesn't yet support this from a client perspective

# 0.0.34 / 2020-07-27

- expose scalar list updates in `0.0.33` to the CLI

# 0.0.33 / 2020-07-27

- add support for all scalar lists including enums (_Postgres only_)

# 0.0.32 / 2020-07-22

- add required has-many support for postgres, add a warning for mysql. Fixes: #38

# 0.0.31 / 2020-07-21

- bump prismafile. fixes: #54.

# 0.0.30 / 2020-07-21

- remove explicit constraint name to allow the db to pick the name. fixes #46
- Add docs link to ID migration section in CLI output

# 0.0.29 / 2020-07-14

- fix column ordering for table relations
- fix indentation for multi-line queries
- fix ids and update tests
- migrate varchar(25) => varchar(30)
- identify ids and columns with foreign keys

# 0.0.28 / 2020-07-13

- fix has-many operation for Postgres & MySQL using example data. Thanks @Aldrian!
- remove datasource from all tests in order to make tests more portable

# 0.0.27 / 2020-07-09

- fix alphanumeric issue with join table

# 0.0.26 / 2020-07-07

- remove console.log

# 0.0.25 / 2020-07-07

- fix has-many SQL statements for DM1.0 (thanks @alan345!)

# 0.0.24 / 2020-07-03

- add support for self-relations

# 0.0.23 / 2020-07-01

- fix 1:N relationships with link:TABLE. add NOT NULL to required 1:1 and 1:N link:TABLE queries. add @db support for the 1:1 and 1:N relationship with link:TABLE
- fix prisma.schema => schema.prisma
- fix message indentation

# 0.0.22 / 2020-06-24

- add tests with both breaking changes and normal changes
- add support for one-to-one relationships in DM1.0 for postgres and mysql
- add breaking ops for 1-to-1 relations in DM1.0
- add support for one-to-many breaking changes. updated the tests
- remove unused functions
- chore: add security email to README

# 0.0.21 / 2020-06-17

- Fix postgres defaults enum (see [note in original commit](https://github.com/prisma/upgrade/commit/c819cfd7d25f43834402d5bd527409b297d1bf03) for caveat message.)
- Cleanup old tests.
- Setup sqldump to work with DM 1.0.
- Add self-relation tests (not passing yet).

# 0.0.20 / 2020-06-12

- add expected.sql to all tests
- fix @createdAt regression
- fix 1-1 unique regression.

# 0.0.19 / 2020-06-09

- ignore 1-to-1 relations that have link: Table

# 0.0.18 / 2020-06-09

- bump prismafile to support comments in P2 schema.
- improve error handling

# 0.0.17 / 2020-06-09

- fix mysql default json.
- improve not null handling in mysql.
- fix cli messaging around enums.

# 0.0.16 / 2020-06-09

- workaround dbgenerated introspection issue for enums with defaults in postgres
- fix mysql5 default text issue
- fix regression where @createdAt getting added superfluously

# 0.0.15 / 2020-06-08

- add support for L: 1-1 relation with both sides required
- add support for @map and @@map

# 0.0.14 / 2020-06-08

- prep for addressing L.
- add test for postgres aliasing
- add K: @db aliasing
- fix tests.
- add better help messages
- make npx the default everywhere

# 0.0.13 / 2020-06-08

- add support for migrating scalar enums. fix missing dependency.
- add missing rollup dependencies
- cleanup unused function

# 0.0.12 / 2020-06-05

- remove @updatedAt setting a default timestamp
- add prettier

# 0.0.11 / 2020-06-05

- cleanup dependencies.
- add ability to skip sql.
- add more commentary.
- add soft clear.

# 0.0.10 / 2020-06-03

- fix version

# 0.0.9 / 2020-06-03

- added a bunch of writing into the tool. uppercase SQL. support env variables inmside prisma.yml
- added multiple datamodel support, support env(...), adjusted signature to prisma.yml schema.prisma, wrapped mysql commands in backticks
- **BREAKING:** We now ask for the `prisma.yml` instead of `datamodel.graphql`

# 0.0.8 / 2020-05-20

- fix schema introspection
- fix: build config

# 0.0.7 / 2020-05-18

- support writing out the final p2 schema

# 0.0.6 / 2020-05-18

- fix @default(cuid()) and @default(uuid())

# 0.0.5 / 2020-05-15

- upgrade prismafile
- fix @updatedAt, remove @default(now())

# 0.0.4 / 2020-05-14

- fix default enums. log the final p2 (temporary)

# 0.0.3 / 2020-05-14

- cleanup dist file

# 0.0.2 / 2020-05-14

- fix console, test blog schema
- mysql 5 test working
- fix up parser and cli
- fix uuid() support
- add prisma 2 printer, support cuid and uuid, support re-introspect, support json
- add initial support for detecting and fixing 1:1 relations with @relation and @relation(link:INLINE)

# 0.0.1 / 2020-04-17

- fix package paths

# 0.0.0 / 2020-04-17

- initial release


src/cli/index.ts
import { bold, red, gray, underline, green } from "kleur"
import { console } from "../console"
import * as prompt from "../prompt"
import * as p2 from "../prisma2"
import * as p1 from "../prisma1"
import * as sql from "../sql"
import * as api from "../api"
// import prompts from 'prompts'
import redent from "redent"
import isURL from "is-url"
import yaml from "js-yaml"
import path from "path"
import util from "util"
import arg from "arg"
import fs from "fs"

/**
 * Constants
 */

const writeFile = util.promisify(fs.writeFile)
const readFile = util.promisify(fs.readFile)
const exists = util.promisify(fs.exists)
const cwd = process.cwd()

/**
 * Flags
 */

const flags = {
  // chdir
  "--chdir": String,
  "-C": "--chdir",

  // url
  "--url": String,
  "-u": "--url",

  // help
  "--help": Boolean,
  "-h": "--help",

  // version
  "--version": Boolean,
  "-v": "--version",
}

function usage() {
  return redent(`
    ${bold("prisma-upgrade")} helps you transition from Prisma 1 to Prisma 2+.

    ${bold("Usage")}

      prisma-upgrade [flags] [path-to-prisma-yml] [path-to-prisma-schema]

    ${bold("Arguments")}

      [path-to-prisma-yml]     Path to prisma.yml file. Defaults to prisma/prisma.yml.
      [path-to-prisma-schema]  Path to schema.prisma file. Defaults to prisma/schema.prisma.

    ${bold("Optional Flags")}

      -u, --url <url>      Connection string to your database.
      -C, --chdir <dir>    Change the working directory.
      -h, --help           Output usage information.
      -v, --version        Show the version.

    ${bold("Examples")}

      ${gray(`# Same as prisma-upgrade prisma/prisma.yml prisma/schema.prisma`)}
      npx prisma-upgrade

      ${gray(`# Specify a custom prisma.yml and schema.prisma path`)}
      npx prisma-upgrade ./prisma.yml ./schema.prisma

      ${gray(`# Specify a Prisma 1 endpoint or connection string`)}
      npx prisma-upgrade --url http://localhost:4467/postgres-env/dev
  `)
}

async function main(argv: string[]): Promise<void> {
  const args = arg(flags, { argv: argv, permissive: true })

  // print help
  if (args["--help"]) {
    console.log(usage())
    return
  }

  // print the version
  if (args["--version"]) {
    console.log(require("../../package.json").version)
    return
  }

  const params = args._.slice(2)
  let prismaYaml = ""
  let schemaPrisma = ""
  switch (params.length) {
    case 0:
      prismaYaml = "prisma/prisma.yml"
      schemaPrisma = "prisma/schema.prisma"
      break
    case 1:
      prismaYaml = params[0]
      schemaPrisma = "prisma/schema.prisma"
      break
    case 2:
      prismaYaml = params[0]
      schemaPrisma = params[1]
      break
    default:
      return fatal(`Too many arguments. Run ${green(`\`npx prisma-upgrade -h\``)} for details.`)
  }

  // change the working directory
  const wd = args["--chdir"] ? path.resolve(cwd, args["--chdir"]) : cwd
  prismaYaml = path.resolve(wd, prismaYaml)
  if (!(await exists(prismaYaml))) {
    return fatal(`prisma.yml doesn't exist in "${prismaYaml}". Run \`prisma-upgrade -h\` for more details.`)
  }
  schemaPrisma = path.resolve(wd, schemaPrisma)
  if (!(await exists(schemaPrisma))) {
    return fatal(`Prisma 2+ schema doesn't exist in "${schemaPrisma}". Run \`prisma-upgrade -h\` for more details.`)
  }

  const yml = yaml.safeLoad(await readFile(prismaYaml, "utf8"))
  if (!yml.endpoint) {
    return fatal(`${bold(`prisma.yml`)} must have an \`endpoint\` parameter`)
  }
  if (!yml.datamodel) {
    return fatal(`${bold(`prisma.yml`)} must have an \`endpoint\` parameter`)
  }

  const datamodel = await concatDatamodels(path.dirname(prismaYaml), yml)
  const prisma1 = p1.parse(datamodel)
  let prisma2: p2.Schema
  try {
    prisma2 = new p2.Schema(await readFile(schemaPrisma, "utf8"))
  } catch (err) {
    const location = err.location ? `${err.location.start.line}:${err.location.start.column} to ${err.location.end.line}:${err.location.end.column}` : ''
    return fatal(
      redent(`
      Error parsing the Prisma 2+ file.

      Are you sure "${schemaPrisma}" is a valid Prisma 2+ schema file?

      Run ${green(`\`prisma-upgrade -h\``)} for more details.

      Stack Trace:

        ${redent(err.stack || err.message, 4).trim()}
      
      Location (line:column):

        ${location}
      `)
    )
  }

  const provider = prisma2.provider()

  // MongoDB is not supported and has a custom upgrade documentation
  if (provider === 'mongodb') {
    return fatal(
      `MongoDB connector is not supported by \`prisma-upgrade\`.`
    )
  }

  const url = isURL(args["--url"] || "") ? args["--url"] : findURL(yml, prisma2)
  if (!url) {
    return fatal(
      redent(`
        We could not determine your endpoint. It's probably in an environment variable.

        Please use the --url flag to specify the endpoint.

        Run ${green(`\`prisma-upgrade -h\``)} for more details.

        Note: The url is not used to connect to your database, only to determine the schema name.
    `)
    )
  }

  // no models
  if (prisma2.models.length === 0) {
    return fatal(
      `Your Prisma 2+ schema doesn't have any models. Run ${green(`\`prisma db pull\``)}, then run ${green(
        `\`prisma-upgrade\``
      )} again.`
    )
  }

  const { ops, schema, breakingOps, idOps } = await api.upgrade({
    prisma1,
    prisma2,
    url,
  })

  if (ops.length || breakingOps.length) {
    console.log(
      redent(`
      ◮ Welcome to the interactive ${bold("Prisma Upgrade CLI")} that helps with the
      upgrade process from Prisma 1 to Prisma 2+.

      Please read the docs to learn more about the upgrade process:
      ${gray(underline("https://pris.ly/d/how-to-upgrade"))}

      ${bold("➤ Goal")}
      The Upgrade CLI helps you resolve the schema incompatibilities
      betweeen Prisma 1 and Prisma 2+. Learn more in the docs:
      ${gray(underline("https://pris.ly/d/schema-incompatibilities"))}

      ${bold("➤ How it works")}
      Troughout the process, you'll need to adjust your database schema by sending
      SQL statements to it. The SQL statements are provided by the Upgrade CLI.

      Note that the Upgrade CLI never makes changes to your database,
      you are in full control over any operations that are executed against it.

      You can stop and re-run the Upgrade CLI at any time.

      These are the different steps of the upgrade processs:

        1. The Upgrade CLI generates SQL commands for you to run on your database.
        2. You run the SQL commands against your database.
        3. You run the ${green(`\`npx prisma db pull\``)} command again.
        4. You run the ${green(`\`npx prisma-upgrade\``)} command again.
        5. Finally the Upgrade CLI adjusts the Prisma 2+ schema by adding missing attributes.

      ${bold("➤ Note")}
      It is recommended that you make a full backup of your existing data before starting
      the upgrade process. If possible, the migration should be performed in a staging
      environment before executed against a production environment.

      ${bold("➤ Help")}
      If you have any questions or run into any problems along the way,
      please create an issue at:
      ${gray(underline("https://github.com/prisma/upgrade/issues/new"))}
    `)
    )
    await confirm(`Are you ready? [Y/n] `)
    clear(true)
  }

  // Non-breaking operations
  const nonBreakingOps = ops.concat(idOps)
  if (nonBreakingOps.length) {
    const segments = new Map<sql.Op["type"], sql.Op[]>()
    for (let op of nonBreakingOps) {
      if (!segments.has(op.type)) {
        segments.set(op.type, [])
      }
      const segment = segments.get(op.type) || []
      segment.push(op)
    }

    console.log(`${bold("➤ Adjust your database schema")}`)
    console.log(`Run the following SQL statements against your database:`)
    console.log()

    // run the queries
    for (let [type, ops] of segments) {
      const queries = sql.translate(provider, ops)
      switch (type) {
        case "SetDefaultOp":
          console.log(`  ${bold(`Add missing \`DEFAULT\` constraints to the database`)}`)
          console.log(
            `  ${gray(`https://pris.ly/d/schema-incompatibilities#default-values-arent-represented-in-database`)}`
          )
          console.log()
          console.log(redent(queries.join("\n"), 4))
          console.log()
          console.log()
          break
        case "SetCreatedAtOp":
          console.log(`  ${bold(`Replicate \`@createdAt\` behavior in Prisma 2+`)}`)
          console.log(`  ${gray(`https://pris.ly/d/schema-incompatibilities#createdat-isnt-represented-in-database`)}`)
          console.log()
          console.log(redent(queries.join("\n"), 4))
          console.log()
          console.log()
          break
        case "AddUniqueConstraintOp":
          console.log(`  ${bold(`Fix 1-1 relations by adding \`UNIQUE\` constraints`)}`)
          console.log(
            `  ${gray(
              `https://pris.ly/d/schema-incompatibilities#inline-1-1-relations-are-recognized-as-1-n-missing-unique-constraint`
            )}`
          )
          console.log()
          console.log(redent(queries.join("\n"), 4))
          console.log()
          console.log()
          break
        case "AlterIDsOp":
          console.log(`  ${bold(`Migrate IDs from varchar(25) to varchar(30)`)}`)
          console.log(`  ${gray(`https://pris.ly/d/schema-incompatibilities#mismatching-cuid-length`)}`)
          console.log()
          console.log(redent(queries.join("\n"), 4))
          console.log()
          console.log()
          break
        case "SetJsonTypeOp":
          console.log(`  ${bold(`Fix columns with JSON data types`)}`)
          console.log(
            `  ${gray(`https://pris.ly/d/schema-incompatibilities#json-type-is-represented-as-text-in-database`)}`
          )
          console.log()
          console.log(redent(queries.join("\n"), 4))
          console.log()
          console.log()
          break
        case "SetEnumTypeOp":
          console.log(`  ${bold(`Fix columns with ENUM data types`)}`)
          console.log(
            `  ${gray(`https://pris.ly/d/schema-incompatibilities#enums-are-represented-as-text-in-database`)}`
          )
          console.log()
          console.log(redent(queries.join("\n"), 4))
          console.log()
          console.log()
          break
      }
    }
  }

  if (breakingOps.length) {
    const segments = new Map<sql.Op["type"], sql.Op[]>()
    for (let op of breakingOps) {
      if (!segments.has(op.type)) {
        segments.set(op.type, [])
      }
      const segment = segments.get(op.type) || []
      segment.push(op)
    }

    console.log(
      redent(`
        ${bold("➤ Breaking changes detected")}

        In order to fully optimize your database schema, you'll need to run a few SQL
        statements that will break your Prisma 1 setup. Note that these changes are optional
        and if you are upgrading gradually and running Prisma 1 and Prisma 2+ side-by-side,
        you should not perform these changes yet. Instead, you can perform them whenever
        you are ready to completely remove Prisma 1 from your project.
        If you are upgrading all at once, you can safely perform these changes now.

        Learn more in the docs:
        ${gray(underline("https://pris.ly/d/how-to-upgrade"))}
      `)
    )
    const yes = await prompt.confirm(`Do you want to view the breaking SQL statements now? [Y/n]? `)
    if (yes) {
      clear(true)

      console.log(`${bold("➤ Adjust your database schema (these changes break Prisma 1)")}`)
      console.log(`Run the following SQL statements against your database:`)
      console.log()

      // run the queries
      for (let [type, ops] of segments) {
        const queries = sql.translate(provider, ops)
        switch (type) {
          case "MigrateHasManyOp":
            console.log(`  ${bold(`Fix one-to-many table relations`)}`)
            console.log(
              `  ${gray(`https://pris.ly/d/schema-incompatibilities#all-non-inline-relations-are-recognized-as-m-n`)}`
            )
            console.log()
            console.log(redent(queries.join("\n"), 4))
            console.log()
            console.log()
            break
          case "MigrateOneToOneTableOp":
            console.log(`  ${bold(`Fix one-to-one table relations`)}`)
            console.log(
              `  ${gray(`https://pris.ly/d/schema-incompatibilities#all-non-inline-relations-are-recognized-as-m-n`)}`
            )
            console.log()
            console.log(redent(queries.join("\n"), 4))
            console.log()
            console.log()
            break
          case "MigrateEnumListOp":
            console.log(`  ${bold(`Fix enum lists`)}`)
            console.log()
            console.log(redent(queries.join("\n"), 4))
            console.log()
            console.log()
            break
          case "MigrateScalarListOp":
            console.log(`  ${bold(`Fix simple scalar lists`)}`)
            console.log()
            console.log(redent(queries.join("\n"), 4))
            console.log()
            console.log()
            break
          case "SetJsonTypeOp":
            console.log(`  ${bold(`Fix columns with JSON data types`)}`)
            console.log(
              `  ${gray(`https://pris.ly/d/schema-incompatibilities#json-type-is-represented-as-text-in-database`)}`
            )
            console.log()
            console.log(redent(queries.join("\n"), 4))
            console.log()
            console.log()
            break
          case "SetEnumTypeOp":
            console.log(`  ${bold(`Fix columns with ENUM data types`)}`)
            console.log(
              `  ${gray(`https://pris.ly/d/schema-incompatibilities#enums-are-represented-as-text-in-database`)}`
            )
            console.log()
            console.log(redent(queries.join("\n"), 4))
            console.log()
            console.log()
            break
        }
      }
    }
  }

  console.log(
    redent(`
      ${bold("➤ Next Steps")}

      After you executed one or more of the previous SQL statements against your database,
      please run the following two commands to refresh your Prisma 2+ schema and check
      the changes.

        1. Run ${green(`\`npx prisma db pull\``)} again to refresh your Prisma 2+ schema.
        2. Run ${green(`\`npx prisma-upgrade\``)} again.

      If you can't or don't want to execute the remaining SQL statements right now, you can
      skip to the last step where the Upgrade CLI adds missing attributes to your Prisma 2+
      schema that are not picked up by introspection.
    `)
  )
  const yes = await prompt.confirm(`Skip to the last step? [Y/n]? `)
  if (!yes) {
    return
  }
  clear(true)

  // overwrite the schema.prisma file if there's no remaining operations
  console.log(
    redent(`
      ${bold("➤ What happens next")}
      As a last step, some final adjustments will be made to your Prisma 2+ schema
      to carry over some Prisma-level attributes that aren't picked up by introspection.

      ${bold("Warning")}
      Your current Prisma 2+ schema will be overwritten, so please
      make sure you have a backup!
  `)
  )
  await confirm(`Are you ready? [Y/n] `)
  clear(true)
  console.log()
  const relpath = path.relative(process.cwd(), schemaPrisma)
  console.log(`Updating ${bold(gray(relpath))}...`)
  await writeFile(schemaPrisma, schema.toString())
  console.log(`Done updating ${bold(gray(relpath))}!`)
  console.log(
    redent(`
      ${bold(`✔ Congratulations, you're all set!`)}

      ${bold("➤ Note")}
      If you didn't execute all generated SQL commands against your database,
      you can re-run the Upgrade CLI at any time.

      Note that the Upgrade CLI doesn't resolve all of the schema incompatibilities
      between Prisma 1 and Prisma 2+. If you want to resolve the remaining ones,
      you can do so manually by following this guide:
      ${gray(underline(`https://pris.ly/d/upgrading-the-prisma-layer`))}

      ${bold("➤ Next steps")}
      Otherwise you can continue your upgrade process by installing Prisma Client 2:
      ${green(`npm install @prisma/client`)}

      You can find guides for different upgrade scenarios in the docs:
      ${gray(underline(`https://pris.ly/d/upgrade-from-prisma-1`))}
  `)
  )
  return
}

/**
 * Concatenate all the datamodels together
 */

async function concatDatamodels(wd: string, yml: any): Promise<string> {
  if (!("datamodel" in yml)) {
    return ""
  }
  const datamodels = [].concat(yml.datamodel)
  const models: string[] = []
  for (let dm of datamodels) {
    models.push(await readFile(path.join(wd, dm), "utf8"))
  }
  return models.join("\n\n")
}

/**
 * Find the URL from prisma.yml or prisma 2 schema
 */

function findURL(yml: any, p2: p2.Schema): string | void {
  if (isURL(yml.endpoint)) {
    return yml.endpoint
  }
  const url = p2.url() || ""
  if (isURL(url)) {
    return url
  }
  return
}

/**
 * Confirm or exit
 */

async function confirm(message: string): Promise<void> {
  const yes = await prompt.confirm(message)
  if (!yes) {
    process.exit(0)
  }
}

/**
 * Fatal message and exit
 */

function fatal(message: string) {
  console.error(
    `\n${red(
      `${bold(`Error`)} ${redent(message).trim()}`
    )}\n\nIf you have any questions or run into any problems along the way, please create an issue at https://github.com/prisma/upgrade/issues/new.\n`
  )
  process.exit(1)
}

/**
 * Clear the console. Optionally maintain scrollback
 */

function clear(isSoft: boolean): void {
  process.stdout.write(isSoft ? "\x1B[H\x1B[2J" : "\x1B[2J\x1B[3J\x1B[H\x1Bc")
}

/**
 * Run main
 */

main(process.argv).catch((err) => {
  fatal(err.message)
  process.exit(1)
})
